### ISSUE TITLE
Add a decimal arithmetics

### ISSUE THREAD

**Issue by XVilka** [ocaml/Zarith#55]
[Decimal floating-point (DFP)](https://en.wikipedia.org/wiki/Decimal_floating_point) arithmetic refers to both a representation and operations on decimal floating-point numbers. Working directly with decimal (base-10) fractions can avoid the rounding errors that otherwise typically occur when converting between decimal fractions (common in human-entered data, such as measurements or financial information) and binary (base-2) fractions.

The advantage of decimal floating-point representation over decimal fixed-point and integer representation is that it supports a much wider range of values. For example, while a fixed-point representation that allocates 8 decimal digits and 2 decimal places can represent the numbers 123456.78, 8765.43, 123.00, and so on, a floating-point representation with 8 decimal digits could also represent 1.2345678, 1234567.8, 0.000012345678, 12345678000000000, and so on. This wider range can dramatically slow the accumulation of rounding errors during successive calculations; for example, the Kahan summation algorithm can be used in floating point to add many numbers with no asymptotic accumulation of rounding error.
Fixed-length decimal point numbers are popular in various business applications, the most notorious example is [COBOL written applications](https://www.ibm.com/support/knowledgecenter/SS6SGM_5.1.0/com.ibm.cobol51.aix.doc/PGandLR/concepts/cpari45.html).

The Rust community for example, has three crates for dealing with decimal numbers: [decimal](https://crates.io/crates/decimal/), [rust-decimal](https://crates.io/crates/rust-decimal), and [bigdecimal](https://crates.io/crates/bigdecimal).

There seems some demand for this type of arithmetic in OCaml world too:
- https://discuss.ocaml.org/t/decimal-floating-point-arithmetic/1447
- https://discuss.ocaml.org/t/exact-rational-numbers/4240

Aside "classic" decimal numbers there are also so called [posits (unums)](https://en.wikipedia.org/wiki/Unum_(number_format)) in OCaml - see [examples in other languages](https://posithub.org/khub_community). There are no benefits of using them while there is no hardware available, nevertheless it might be interesting to play with the software implementation for evaluation reasons.

See also [Decimal Point Algorithms for Computers](http://speleotrove.com/decimal/IEEE-cowlishaw-arith16.pdf) and [Software Implementation of DecimalFloating-Point](https://www.cl.cam.ac.uk/~jrh13/slides/jnao-01jun06/slides.pdf) papers.

**Comment by pascal-cuoq** [comment]
Anyone who is interested in decimal floating-point in OCaml can interface Intel's software implementation of IEEE 754's decimal{32,64,128} formats. This would be much better than inventing other subpar formats as each framework seems to have been doing until now. This implementation is described at https://software.intel.com/en-us/articles/intel-decimal-floating-point-math-library and corporate websites being what they are, I should point out that the most important information on the page is the link to http://www.netlib.org/misc/intel/ which is likely to be more stable in time.

Such bindings, however, would not share a single line of code with the current Zarith, and since floating-point on the one hand and bignum/rationals on the other hand are alternative solutions to some of the same problems, most users of a new Zarith that included Intel's decimal FP implementation would only take advantage of half the library (and half of the dependencies; the other dependencies would be bloat).

TLDR: you are asking for a whole new library, and there is no obvious benefit to making this new library part of Zarith.

**Comment by antoinemine** [comment]
I agree that this feature is out of scope for ZArith, which targets only integer (mainly) and rational arithmetic (to a lesser extent). Its features are based on those in GMP (minus the float, which is binary anyway). As @pascal-cuoq  suggests, it would be cleaner to make a separate library. It could either be a binding to an existing decimal float library, or a full OCaml re-implementation (possibly based on ZArith). Sorry...

**Comment by yawaramin** [comment]
FYI, I followed the second suggestion: http://opam.ocaml.org/packages/decimal/

**Comment by antoinemine** [comment]
@yawaramin thanks for the link, this looks quite interesting!
